runner.py

```diff
@@ -0,0 +1,285 @@
+            stdout, stderr = process.communicate()
+            process.wait()
+
+            # Called only after the command has completed.
+            self.__executed(process.returncode, stdout, stderr)
+
+    def __executed(self, code, stdout, stderr):
+        """Called when the command has finished executing.
+
+        Parameters
+        ----------
+        code : bool
+            The return code of the command.
+        stdout : ellipsis
+            Standard command output.
+        stderr : ellipsis
+            Standard command error.
+        """
+        logging.debug(f'__on_executed() -> return_code: {code}')
+
+        # If the return code is not 0, then it failed.
+        if code != 0:
+            self.attempts -= 1
+
+            if self.__sys_trace:
+                create_log_file('sys-trace', self.__captured_sys_trace)
+
+            if self.__call_trace:
+                create_log_file('call-trace', f'\n{stderr.decode()}')
+
+            # TODO: click.echo() should not be in the command class.
+            if self.__log_trace:
+                if stdout:
+                    click.echo(f'\n{stdout.decode()}')
+
+                if stderr:
+                    click.echo(f'\n{stderr.decode()}')
+
+        # Add or update the return code for the execution summary.
+        self.__return_codes[code] = self.__return_codes.get(code, 0) + 1
+
+    @property
+    def attempts(self):
+        """Returns the attempts."""
+        logging.debug('attempts.getter')
+        return self.__attempts
+
+    @attempts.setter
+    def attempts(self, value):
+        """Set the attempts value.
+
+        Parameters
+        ----------
+        value : int
+            The new value.
+        """
+        logging.debug('attempts.setter')
+        self.__attempts = value
+
+    def summary(self):
+        """Returns a summary of the command execution."""
+        logging.debug('summary()')
+
+        result = '--- command execution statistics ---'
+        codes = self.__return_codes
+
+        # If no return codes are available, return the result.
+        if len(codes) == 0:
+            logging.debug('summary() -> len(__return_codes) = 0')
+            return result
+
+        # Add all return codes to the result.
+        for key, value in codes.items():
+            result += (f'\nreturn code: {key} amount: {value}')
+
+        # Add the most frequent return code to the result.
+        result += f'\nmost frequent return code: {max(codes, key=codes.get)}'
+        return result
+
+
+def capture_sys_trace(process_id):
+    """Captures and returns the system trace for the specified process.
+
+    Parameters
+    ----------
+    process_id : int
+        The process identifier.
+    """
+    logging.debug('capture_sys_trace()')
+
+    # Captures the specified process.
+    process = psutil.Process(process_id)
+
+    # Receives a trace and returns it.
+    return [f'Disk I/O: {str(psutil.disk_io_counters(perdisk=False))}',
```

Comment:
you should return a dictionary instead, and format it outside somewhere.

the method should only capture the systrace into a proper data structure.
this list of strings that is returned is only good for printing and nothing else.